DNN Platform/Library/ExtensionPoints/SafeDirectoryCatalog.cs

```diff
@@ -34,19 +35,51 @@ public class SafeDirectoryCatalog : ComposablePartCatalog
 
         public SafeDirectoryCatalog(string directory)
         {
-            var files = Directory.EnumerateFiles(directory, "*.dll", SearchOption.AllDirectories);
-
             _catalog = new AggregateCatalog();
 
+            // ASP.NET will load all the DLL files under the bin folder; the first loop would be sufficient
+            var loadedAssemblies = AppDomain.CurrentDomain.GetAssemblies()
+                                            .Where(asm => asm.GetName().CodeBase.StartsWith("file://"));
```

Comment:
This is somewhat personal preference, but I'd prefer to see more of the conditions moved into the LINQ statement, and out of the loop (see http://www.martinfowler.com/articles/refactoring-pipelines.html).  Something like this, perhaps:

```
var matchingAssemblyFiles = (from assembly in AppDomain.CurrentDomain.GetAssemblies()
                             let codeBase = assembly.GetName().CodeBase
                             where codeBase.StartsWith("file://")
                             let fileUri = new Uri(codeBase)
                             let fileInfo = new FileInfo(fileUri.AbsolutePath)
                             where fileInfo.DirectoryName != null
                             where fileInfo.DirectoryName.StartsWith(directory, StringComparison.OrdinalIgnoreCase)
                             select new {
                                 fileInfo.FullName,
                                 Assembly = assembly,
                             }).ToList();

foreach (var assembly in matchingAssemblyFiles.Select(f => f.Assembly)) {
    TryAddCatalog(_catalog, assembly);
}

var otherAssemblyFiles = Directory.EnumerateFiles(directory, "*.dll", SearchOption.AllDirectories)
                                  .Except(matchingAssemblyFiles.Select (f => f.FullName), StringComparer.OrdinalIgnoreCase);
                                  
foreach (var assemblyPath in otherAssemblyFiles) {
    TryAddCatalog(_catalog, assemblyPath);
}
```